interfaces/cos_agent/v0/schema.py

```diff
@@ -0,0 +1,140 @@
+                ],
+                "log_slots": ["charmed-zookeeper:logs"]
+            }
+    RequirerSchema:
+        unit: <empty>
+        app: <empty>
+"""
+
+import base64
+import json
+import lzma
+from pydantic import BaseModel
+from typing import ClassVar, Dict, List, Union, Optional
+from interface_tester.schema_base import DataBagSchema
+
+
+class GrafanaDashboard(str):
+    """Grafana Dashboard encoded json; lzma-compressed."""
+
+    # TODO Replace this with a custom type when pydantic v2 released (end of 2023 Q1?)
+    # https://github.com/pydantic/pydantic/issues/4887
+    @staticmethod
+    def _serialize(raw_json: Union[str, bytes]) -> "GrafanaDashboard":
+        if not isinstance(raw_json, bytes):
+            raw_json = raw_json.encode("utf-8")
+        encoded = base64.b64encode(lzma.compress(raw_json)).decode("utf-8")
+        return GrafanaDashboard(encoded)
+
+    def _deserialize(self) -> Dict:
+        raw = lzma.decompress(base64.b64decode(self.encode("utf-8"))).decode()
+        return json.loads(raw)
+
+    def __repr__(self):
+        """Return string representation of self."""
+        return "<GrafanaDashboard>"
+
+
+class CosAgentProviderUnitData(BaseModel):
+    """Unit databag model for `cos-agent` relation."""
+
+    # The following entries are the same for all units of the same principal.
+    # Note that the same grafana agent subordinate may be related to several apps.
+    # this needs to make its way to the gagent leader
+    metrics_alert_rules: dict
+    log_alert_rules: dict
+    dashboards: List[GrafanaDashboard]
+
+    # The following entries may vary across units of the same principal app.
+    # this data does not need to be forwarded to the gagent leader
+    metrics_scrape_jobs: List[Dict]
+    log_slots: List[str]
+
+    # when this whole datastructure is dumped into a databag, it will be nested under this key.
+    # while not strictly necessary (we could have it 'flattened out' into the databag),
+    # this simplifies working with the model.
+    KEY: ClassVar[str] = "config"
+
+
+class CosAgentPeersUnitData(BaseModel):
+    """Unit databag model for `cluster` cos-agent machine charm peer relation."""
+
+    # We need the principal unit name and relation metadata to be able to render identifiers
+    # (e.g. topology) on the leader side, after all the data moves into peer data (the grafana
+    # agent leader can only see its own principal, because it is a subordinate charm).
+    principal_unit_name: str
+    principal_relation_id: str
+    principal_relation_name: str
+
+    # The only data that is forwarded to the leader is data that needs to go into the app databags
+    # of the outgoing o11y relations.
+    metrics_alert_rules: Optional[dict]
+    log_alert_rules: Optional[dict]
+    dashboards: Optional[List[GrafanaDashboard]]
+
+    # when this whole datastructure is dumped into a databag, it will be nested under this key.
```

Comment:
the pydantic schemas in this file should reflect the ACTUAL shape of the data structure, not the way we have implemented it. 
So the KEY should go from both models, and instead there should be something like:

```
class MyAppDataModel(BaseModel):
     config: Json[MyNestedDataModel]  # if json-encoded. If it's yaml, you need to add a validator like I did with ingress.
     
class MyNestedDataModel(BaseModel):
    metrics_alert_rules = ... 
    log_alert_rules = ...
```